backend/ml/tests/test_mehestan_primitives.py

```diff
@@ -0,0 +1,66 @@
+import numpy as np
+from django.test import TestCase
+
+from ml.mehestan.primitives import QrMed, QrUnc
+
+
+class QrMedTest(TestCase):
+    def test_qrmed(self):
+        W = 2
+        weight = 1
+        self.assertAlmostEqual(
+            QrMed(
+                W,
+                w=weight,
+                x=np.array([-10.0, 1.0, 10.0]),
+                delta=np.array([1e-3, 1e-3, 1e-3]),
+            ),
+            0.5, # influence of the median contributor is bounded by weight / W
+            places=3
+        )
+
+    def test_qrmed_with_W_equals_zero(self):
+        W = 0
+        weight = 1
+        self.assertAlmostEqual(
+            QrMed(
+                W,
+                w=weight,
+                x=np.array([-10.0, 1.0, 10.0]),
+                delta=np.array([1e-3, 1e-3, 1e-3]),
+            ),
+            1.0,
+            places=3
+        )
+
+
+class QrUncTest(TestCase):
+    def test_qrunc(self):
+        W = 2
+        weight = 1
+        self.assertAlmostEqual(
+            QrUnc(
+                W,
+                default_dev=0,
+                w=weight,
+                x=np.array([-10.0, 1.0, 10.0]),
+                delta=np.array([1e-3, 1e-3, 1e-3]),
+            ),
+            0.506,
+            places=3,
+        )
+
+    def test_qrunc_with_W_equals_zero(self):
+        W = 0
+        weight = 1
+        self.assertAlmostEqual(
+            QrUnc(
+                W,
+                default_dev=0,
+                w=weight,
+                x=np.array([-10.0, 1.0, 10.0]),
+                delta=np.array([1e-3, 1e-3, 1e-3]),
+            ),
+            0.033,
```

Comment:
I'm not sure how to interpret the fact that QrUnc is much lower when W is equal to 0.

Reply:
@lenhoanglnh QrUnc looks like it gives weird results depending on W (other variables are here the same as in Adrien's test). I don't think the problem comes from the python calculation, I think it's the formula that is not perfect.

![image](https://user-images.githubusercontent.com/47578089/198912360-2e53f3fd-ef06-49e4-b174-92320adf72a4.png)

Reply:
Here is when W = 2 and the weights vary :
![image](https://user-images.githubusercontent.com/47578089/198912999-29a4a5c0-e3b7-4127-9545-5d7967299639.png)

Here is when W = 2 and each score is increased by x (scores = [-10 + x, 1 + x, 10 + x]) :
![image](https://user-images.githubusercontent.com/47578089/198913373-77245fa8-509a-4bb3-be32-cd09ff77eee6.png)

This is very weird

Reply:
and for the uncertainty

![image](https://user-images.githubusercontent.com/47578089/198914560-343f83b5-0d34-4130-b29c-8736a74b5079.png)

Reply:
QrUnc(default_dev) looks almost good

![image](https://user-images.githubusercontent.com/47578089/199100888-83e032e9-8cfe-4e9a-8d7f-8eb3cb8c48dc.png)

Reply:
Sorry Adrien, I forgot this PR. What you made looks correct to me. But the behaviour of QrUnc looks very weird and I think it's well worth investigating, notably whether you find the same patterns of curves. Not necessarily in this PR though, so I can validate if you want to do it separately.

Reply:
I'll investigate it in-depth this week, please let me a few days.

Reply:
I continue to investigate, but one thing that may be interesting would be to use the LogSumExp smooth minimum instead of the SoftMinMean.

One reason is that the SoftMinMean isn't really consistent, you can have x1 > x2 and yet SoftMinMean(x1, y) < SoftMinMean(x2, y). Moreover, (I didn't understand why, but) in the paper we say that both elements of SoftMeanMin "represent upper bounds on uncertainty" obtained with SoftMeanMin. So we want SoftMeanMin(x, y) < x and SoftMeanMin(x, y) < y. But this is not true, as shown in the following figure, it's the contrary.

The magenta curve represents the min and the red curves the SoftMinMean.

![image](https://user-images.githubusercontent.com/47578089/206081514-05c7aca9-7a2a-498e-8392-76d9430d24ec.png)

The LogSumExp looks like it fits more what we are trying to do :

![image](https://user-images.githubusercontent.com/47578089/206081715-d2bfcd7b-aeda-4950-8e62-175a0af9b104.png)

Here is the code to generate and modify the images :
https://colab.research.google.com/drive/1QcNQXFxGcu9PnLPIhyGafCH6QyNx936t?usp=sharing

Reply:
That's very nice you can look deeper into this. I'm not really familiar with the initial motivation for using SoftminMax, but I would be curious to see the impact on "QrUnc" if LogSumExp would be use instead.

What do you think of opening a new issue to track the potential changes to apply on `QrUnc` and discuss this?

If there is no objection, I would like to merge this PR as it is. Even if it does not solve everything, it would bring consistency between the paper and what is used on production, and that would useful as a baseline. What do you think?

Reply:
Sure. And I'm starting to think that the fact that the curves are weird probably doesn't reveal a problem. I haven't finished to investigate, but it looks like when there is too little voting rights, there can be some clipping that causes steep irregularities.

I still think using LogSumExp is an improvement and opened the issue #1232 , but it shouldn't have a major effect on the weirdness of the curves.